FirebaseAdmin/FirebaseAdmin/Messaging/WebpushNotification.cs

```diff
@@ -0,0 +1,276 @@
+// Copyright 2018, Google Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Google.Apis.Json;
+using Newtonsoft.Json;
+
+namespace FirebaseAdmin.Messaging
+{
+    /// <summary>
+    /// Represents the Webpush-specific notification options that can be included in a
+    /// <see cref="Message"/>. Supports most standard options defined in the
+    /// <see href="https://developer.mozilla.org/en-US/docs/Web/API/notification/Notification">
+    /// Web Notification specification</see>
+    /// </summary>
+    public sealed class WebpushNotification
+    {
+        /// <summary>
+        /// Title text of the notification.
+        /// </summary>
+        [JsonProperty("title")]
+        public string Title { get; set; }
+
+        /// <summary>
+        /// Body text of the notification.
+        /// </summary>
+        [JsonProperty("body")]
+        public string Body { get; set; }
+
+        /// <summary>
+        /// The URL to the icon of the notification.
+        /// </summary>
+        [JsonProperty("icon")]
+        public string Icon { get; set; }
+
+        /// <summary>
+        /// The URL of the image used to represent the notification when there is not enough space
+        /// to display the notification itself.
+        /// </summary>
+        [JsonProperty("badge")]
+        public string Badge { get; set; }
+
+        /// <summary>
+        /// Any arbitrary data that should be associated with the notification.
+        /// </summary>
+        [JsonProperty("data")]
+        public object Data { get; set; }
+
+        /// <summary>
+        /// The direction in which to display the notification.
+        /// </summary>
+        [JsonIgnore]
+        public Direction? Direction { get; set; }
```

Comment:
I'm wondering why you have a set of nullable properties for types like this?  Why not use default values instead?  To me it seems like a pain to deal with the nullability of e.g bool properties rather than just reading the default value.

Reply:
This is to make sure those parameters get dropped from the JSON output. For most parameters, we want to avoid sending a default value to the backend (0 for ints, false for bools etc) when they are not set by the user.

Reply:
It does worry me a little that we're letting optimizations leak into the API surface.  I guess if that's the way you do it across the board it's ok to me.